lib/puppet-lint/data.rb

```diff
@@ -155,6 +155,21 @@ def title_tokens
       end
     end
 
+    # Internal: Determine if the given token contains a CLASSREF in
+    # the token chain..
+    #
+    # Returns a Boolean.
+    def classref?(token)
+      current_token = token
+      while true # rubocop:disable Style/InfiniteLoop
+        current_token = current_token.prev_code_token
+        break if current_token.nil?
+
+        return true if current_token.type == :CLASSREF
+        return false if current_token.type == :NAME
```

Comment:
Looking at the code, doesn't this backtrack all the way until it finds something that's a classref or name? For example, what if you nest conditional statements?

I'm just wondering if syntactically colon is always preceded by a CLASSREF or NAME and if there's nothing in between.

Reply:
I really wondered, but couldn't quickly think of an example. After a quick look through [the grammar](https://github.com/puppetlabs/puppet/blob/main/lib/puppet/pops/parser/egrammar.ra) it looks like case statements also contain colons. So maybe something like:
```puppet
service { 'foo': }

case $var {
  'value': {}
}
```

And some variations of that.

Reply:
I think in this case we are safe because resource_indexes will skip the given token if it contains `:LBRACE`.

`next unless colon_token.next_code_token && colon_token.next_code_token.type != :LBRACE`

However I do wonder if we could potentially save cycles if we moved this new check so it happens after the above clause...

Reply:
Latest push has the above change